tmt/utils.py

```diff
@@ -6120,30 +6120,60 @@ def regex_search(
     return environment
 
 
-def render_template_file(
-        template_filepath: Path,
+def render_template(
+        template: str,
+        template_filepath: Optional[Path] = None,
         environment: Optional[jinja2.Environment] = None,
         **variables: Any
         ) -> str:
-    """ Render a template read from a file """
+    """
+    Render a template.
+
+    :param template: template to render.
+    :param template_filepath: path to the template file, if any.
+    :param environment: Jinja2 environment to use.
+    :param variables: variables to pass to the template.
+    """
 
     environment = environment or default_template_environment()
 
     try:
-        template = environment.from_string(template_filepath.read_text())
-
-        return template.render(**variables).strip()
-
-    except FileNotFoundError as exc:
-        raise GeneralError(
-            f"Could not open template '{template_filepath}'.") from exc
+        return environment.from_string(template).render(**variables).strip() + '\n'
```

Comment:
Why are we adding the extra `\n`?

Reply:
We need to make sure that there is a newline at the end of the template string, because users may add additional information to it. For instance, if the `--link` option is specified, the link would be appended to the last line instead of on a separate line.

Reply:
Thanks for clarification. Thinking a bit more: So... wouldn't it be better to remove the whole `.strip() + '\n'` part? Why do we strip the template and then add the new line back? @happz, any idea why the `strip()` part was included in afc12dccc?

Reply:
Oh, I missed this one. An important one.

`strip()` makes composing multiple rendered templates easier to reason about. It does not assume how they should be put together. Whatever comes out of `render_template()` has no leading whitespace, and no trailing whitespace. If the caller of `render_template()` wants whitespace to be there, three empty lines or seven tabs plus a carriage return, it's their responsibility to add it. This is exactly what happens in https://github.com/teemtee/tmt/commit/afc12dccc74e5368e897519a057ed3a42bfb1fd1#diff-85933aa74a2d66c3e4dcdf7a9ad8397f5a7971080d34ef1108296a7c6b69e7e3R287 or https://github.com/teemtee/tmt/commit/afc12dccc74e5368e897519a057ed3a42bfb1fd1#diff-c52a6536c4e908bd39490454987527e616ba4038032346fea86f449940098910R29.

Adding `\n` does not belong to `render_template()`, the template manager should add it to rendered templates if it's needed for that particular kind of output.

Reply:
Thank you for the clarification. I moved the newline from `render_template()` to template manager in 00a55196bc1ece3b068d746aed56aaafd844f88c.

Reply:
Good! @happz, thanks for the great clarification and @therazix for fixing it!